File: ProjectEuler/051-100/S77.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ProjectEuler
{
    public class S77 : ISolution
    {
        private static readonly int Max = 5000;

        public string GetAnswer()
        {
            var primes = Utils.GenPrimes(Max).Where(l => l != 0).Reverse().Select(Convert.ToInt32).ToArray();
            for (int i = 2; i < Max; i++)
            {
                int count = Count(i, primes);
                if (count >= Max)
                {
                    return i.ToString();
                }
            }

            return string.Empty;
        }

        private static int Count(int sum, int[] primes)
        {
            int index = Array.BinarySearch(primes, sum);
            if (index < 0)
            {
                index = ~index;
            }

            return Count(sum, index, primes);
        }

        private static int Count(int sum, int index, int[] primes)
        {
            if (sum == 0)
            {
                return 1;
            }

            if (sum < 0 || index >= primes.Length)
            {
                return 0;
            }

            return Count(sum - primes[index], index, primes) + Count(sum, index + 1, primes);
        }
    }
}

File: src/051-100/S77.cc
#include "S77.h"
#include "../Utils.h"
#include <iostream>
#include <string>
#include <vector>

using namespace std;

namespace ProjectEuler
{
std::string S77::GetAnswer()
{
	int limit = 1000;
	auto sieve = Utils::GenPrimes(limit);
	std::vector<int> primes;
	for (auto p : sieve)
	{
		if (p != 0)
		{
			primes.push_back((int)p);
		}
	}

	std::vector<int> ways(limit + 1, 0);
	ways[0] = 1;

	for (int p : primes)
	{
		for (int i = p; i <= limit; i++)
		{
			ways[i] += ways[i - p];
		}
	}

	for (int i = 0; i <= limit; i++)
	{
		if (ways[i] > 5000)
		{
			return std::to_string(i);
		}
	}

	return "0";
}
} // namespace ProjectEuler
